File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/object/UaBrowseObjectTransaction.java
package org.opcfoundation.webserver.service.transactions.object;

import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.QualifiedName;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription;
import org.eclipse.milo.opcua.stack.core.types.structured.ReferenceDescription;
import org.opcfoundation.webserver.addressspace.models.UaObjectType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManager;
import org.opcfoundation.webserver.service.transactions.base.UaBrowseTransaction;
import org.opcfoundation.webserver.types.*;
import org.opcfoundation.webserver.types.message.BrowseObjectRequest;
import org.opcfoundation.webserver.types.message.BrowseObjectResponse;
import org.opcfoundation.webapi.service.types.ServiceContext;

import java.util.concurrent.CompletableFuture;

public class UaBrowseObjectTransaction extends UaBrowseTransaction {
    private final UaObjectType objectType;
    private final UaObjectId objectId;
    private final NodeManager nodeManager;

    public UaBrowseObjectTransaction(
            ServiceContext serviceContext,
            BrowseDescription nodeToBrowse,
            UaBrowseAdditionalInfo additionalInfo,
            int handleId,
            UaObjectType objectType,
            UaObjectId objectId,
            NodeManager nodeManager)
    {
        super(serviceContext, nodeToBrowse, additionalInfo, handleId);
        this.objectType = objectType;
        this.objectId = objectId;
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute() {
        try {
            // Add type definition
            if (additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_DEFINITION_TASK))
            {
                addTypeDefinitionReference(
                        objectType.nodeId(),
                        objectType.browseName(),
                        objectType.displayName());

                additionalInfo = additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_DEFINITION_TASK);
            }

//            if (!objectType.isGetParentSupported())
//            {
                additionalInfo = additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_PARENT_TASK);
//            }

            if (!objectType.isGetLinkSupported())
            {
                additionalInfo = additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_LINK_TASK);
            }

            if (additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK) ||
                    additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK) ||
                    additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK)) {
                BrowseObjectRequest request = new BrowseObjectRequest(
                        objectId,
                        additionalInfo,
                        getItem());

                return objectType.onBrowseObjectChildren(request).
                        thenApply(this::browseObjectResult).
                        exceptionally(ex -> buildErrorResponse(ex.getCause()));
            }

            if (additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_LINK_TASK)) {
                BrowseObjectRequest request = new BrowseObjectRequest(
                        objectId,
                        additionalInfo,
                        getItem());

                return objectType.onBrowseObjectLinks(request).
                        thenApply(this::browseObjectResult).
                        exceptionally(ex -> buildErrorResponse(ex.getCause()));
            }

            /*
            if (additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_PARENT_TASK)) {
                BrowseObjectRequest request = new BrowseObjectRequest(
                        objectId,
                        additionalInfo,
                        getItem());

                return objectType.onBrowseObjectParent(request).
                        thenApply(response-> browseObjectResult(response, UaBrowseAdditionalInfo.GET_PARENT_TASK)).
                        exceptionally(ex -> buildErrorResponse(ex.getCause()));
            } */
        } catch (Exception e) {
            buildErrorResponse(e);
        }

        return CompletableFuture.completedFuture(null);
    }

    private Void browseObjectResult(BrowseObjectResponse response)
    {
        for (UaReferenceDescriptor item: response.getChildren())
        {
            if (item.getId().isEmpty()) continue;

            if (item.getNodeClass() != NodeClass.Object &&
                    item.getNodeClass() != NodeClass.Variable &&
                    item.getNodeClass() != NodeClass.Method) continue;

            UaInstanceIdentifier newIdentifier;

            if (item.getNodeClass() == NodeClass.Object)
            {
                UaObjectIdentifier objectIdentifier = new UaObjectIdentifier(
                        item.getTypeDefinitionId().toParseableString(),
                        item.getId(),
                        (item.getInstanceDeclarationId().isNull()) ? null : item.getInstanceDeclarationId().toParseableString());

                newIdentifier = new UaInstanceIdentifier(
                        objectIdentifier,
                        null);
            } else {
                UaObjectIdentifier objectIdentifier = new UaObjectIdentifier(
                        objectType.nodeId().toParseableString(),
                        objectId.getId(),
                        (null == objectId.getInstance()) ? null : objectId.getInstance().nodeId().toParseableString());

                UaChildIdentifier memberIdentifier = new UaChildIdentifier(
                        item.getId(),
                        null,
                        item.getNodeClass() == NodeClass.Method);

                newIdentifier = new UaInstanceIdentifier(
                        objectIdentifier,
                        memberIdentifier);
            }

            references.add(new ReferenceDescription(
                    item.getReferenceTypeId(),
                    item.isForward(),
                    new NodeId(nodeManager.nsIndex(), newIdentifier.toByteString()).expanded(),
                    new QualifiedName(0,item.getBrowseName()),
                    item.getDisplayName(),
                    item.getNodeClass(),
                    item.getTypeDefinitionId().expanded()));
        }

        if (response.containsMoreData() && !references.isEmpty())
        {
            additionalInfo = additionalInfo.updateOffset(references.size());
        } else {
            additionalInfo = additionalInfo.taskComplete(response.getTaskMask());
        }

        if (!additionalInfo.isAllTaskComplete())
        {
            continuationPoint = new UaBrowseContinuationPoint(getItem(), additionalInfo).toByteString();
        }

        return null;
    }
}

File: opcua-webservice-node/source/src/server/service/transactions/reactiveobject/UaBrowseObjectTransaction.ts
import { NodeClass } from "opcua-webapi";
import { UaBrowseDescription, UaExpandedNodeId, UaReferenceDescription, UaNodeId, UaNodeIdType } from "opcua-webapi-ts";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaReactiveObjectType } from "../../../addressspace/reactiveobject/UaReactiveObjectType";
import {
    UaBrowseAdditionalInfo,
    UaBrowseContinuationPoint,
    UaChildIdentifier,
    UaInstanceIdentifier,
    UaObjectId,
    UaObjectIdentifier,
} from "../../../types";
import { BrowseObjectRequest, BrowseObjectResponse } from "../../message";
import { UaBrowseTransaction } from "../base/UaBrowseTransaction";
import { ServiceContext, UaReferenceType } from "../../..";

export class UaBrowseObjectTransaction extends UaBrowseTransaction {
    private readonly objectType: UaReactiveObjectType;
    private readonly objectId: UaObjectId;
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(
        serviceContext: ServiceContext,
        nodeToBrowse: UaBrowseDescription,
        additionalInfo: UaBrowseAdditionalInfo,
        handleId: number,
        objectType: UaReactiveObjectType,
        objectId: UaObjectId,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(serviceContext, nodeToBrowse, additionalInfo, handleId);
        this.objectType = objectType;
        this.objectId = objectId;
        this.nodeManager = nodeManager;
    }

    override async execute(): Promise<void> {
        try {
            const request = new BrowseObjectRequest(
                this.objectId,
                this._additionalInfo,
                this.getItem(),
            );

            if (this._additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_DEFINITION_TASK)) {
                this.addTypeDefinitionReference(
                    this.objectType.nodeId,
                    this.objectType.browseName,
                    this.objectType.displayName);
                this._additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_DEFINITION_TASK);
            }

            let referenceType = this.nodeManager.getNode(this.objectType.supportedReferenceType());
            let includeSubtypes = this.getItem().includeSubtypes;

            if (referenceType === null || referenceType.nodeClass != NodeClass.ReferenceType)
            {
                this._additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK);
            } else {
                if (includeSubtypes) {
                    if (!(referenceType as UaReferenceType).isSubtypeOf(this.getItem().referenceTypeId))
                    {
                        this._additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK);
                    }
                } else {
                    if (!referenceType.nodeId.equal(this.getItem().referenceTypeId))
                    {
                        this._additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK);
                    }
                }                
            }

            if (this._additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK) ||
                this._additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK) ||
                this._additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK))
            {
                const response = await this.objectType.onBrowseObject(request);
                this.browseObjectResult(response);
            }
        } catch (error) {
            this.buildErrorResults(error);
        }
    }

    private browseObjectResult(response: BrowseObjectResponse)
    {
        for (const item of response.children)
        {
            if (item.nodeClass == NodeClass.ObjectType)
            {
                this._references.push(new UaReferenceDescription(
                        UaExpandedNodeId.from(item.typeDefinitionId),
                        item.nodeClass,
                        item.browseName,
                        item.displayName,
                        item.referenceTypeId,
                        true,
                        UaExpandedNodeId.from(UaNodeId.nullNodeId)));

                continue;
            }

            if (item.id.length === 0) continue;

            if (item.nodeClass != NodeClass.Object &&
                    item.nodeClass != NodeClass.Variable &&
                    item.nodeClass != NodeClass.Method) continue;

            let newIdentifier : UaInstanceIdentifier;

            if (item.nodeClass == NodeClass.Object)
            {
                let objectIdentifier = new UaObjectIdentifier(
                        item.typeDefinitionId.toString(),
                        item.id,
                        (!item.instanceDeclarationId.isEmpty()) ? item.instanceDeclarationId.toString() : null);

                newIdentifier = new UaInstanceIdentifier(
                        objectIdentifier,
                        null);
            } else {
                let objectIdentifier = new UaObjectIdentifier(
                        this.objectType.nodeId.toString(),
                        this.objectId.id,
                        (this.objectId.instance) ? this.objectId.instance.nodeId.toString() : null);

                let memberIdentifier = new UaChildIdentifier(
                        item.id,
                        null);

                newIdentifier = new UaInstanceIdentifier(
                        objectIdentifier,
                        memberIdentifier);
            }

            let newNodeId = new UaNodeId(
                newIdentifier.toByteString(),
                this.nodeManager.nsIndex(),
                UaNodeIdType.BYTESTRING);
                
            this._references.push(new UaReferenceDescription(
                    UaExpandedNodeId.from(newNodeId),
                    item.nodeClass,
                    item.browseName,
                    item.displayName,
                    item.referenceTypeId,
                    true,
                    UaExpandedNodeId.from(item.typeDefinitionId)));
        }

        if (0 != response.remainingTasks && 0 != this._references.length)
        {
            this._additionalInfo.taskCheckListMasks = response.remainingTasks;
            this._additionalInfo.referenceOffset = response.offset;
        } else {
            this._additionalInfo.taskCheckListMasks = 0;
            this._additionalInfo.referenceOffset = 0;
        }

        if (!this._additionalInfo.isAllTaskComplete())
        {
            this._continuationPoint = new UaBrowseContinuationPoint(this.getItem(), this._additionalInfo).toByteString();
        }

        return null;
    }
}
